debug ival swap

avikchaudhuri opened the pull request:
Summary:
Recall that we use "ivals" to track intermediate values of mutations during unflattening. Previously, for each such intermediate value, we would create a hidden shared attribute that would be updated / read by respective submodules.

Unfortunately this scheme doesn't work when some but not all of those submodules are swapped out. This is because the swapped in submodules have no knowledge of these hidden attributes. Thus the submodules that are not swapped out end up reading / updating dangling state.

This PR does away with these hidden attributes. Instead, we directly read the underlying buffer or placeholder that was updated, and update those underlying buffers and placeholders in place. This makes the graphs look much closer to their eager origins.

Test Plan: added some tests, ensured existing tests pass

Differential Revision: D71203469

tugsbayasgalan commented on torch/export/unflatten.py:
```diff
@@ -711,62 +709,74 @@ def unflatten(
+    buffer_fqn_to_input_name = {
+        buffer_fqn: k for k, buffer_fqn in graph_signature.inputs_to_buffers.items()
+    }
+    for k, buffer_fqn in graph_signature.buffers_to_mutate.items():
```
In training IR, this signature is empty.

tugsbayasgalan replied:
Or i guess in training IR, we would have returned earlier?

avikchaudhuri replied:
no problem with training IR because the mutations are already in the graph

tugsbayasgalan commented on torch/export/unflatten.py:
```diff
@@ -1423,79 +1430,42 @@ class _IVals:
+        self.node_names = set()
 
-        return ival_node
+    def _is_mutable(self, target):
```
There are some ops that are maybe_aliasing or aliasing that you should extra check.

avikchaudhuri replied:
I checked that `.to` is still a mutation. Test coming up...

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -4736,6 +4736,138 @@ def forward(self, x, y):
         self.assertTrue(torch.allclose(orig_res[1], ep_res[1]))
         self.assertTrue(torch.allclose(orig_res[2], ep_res[2]))
 
+    def test_unflatten_placeholder_update_child2parent_swap(self):
```
Can you add a test case dropout where it is forced to alias (in inference case this will happen)? This will make sure our mutation detecting thing is working fine.

avikchaudhuri replied:
done

tugsbayasgalan commented on torch/export/unflatten.py:
```diff
@@ -1423,78 +1430,51 @@ class _IVals:
+                    with graph.inserting_after(node):
+                        new_node = graph.create_node(
+                            "call_function",
+                            torch.ops.aten.copy_,
```
should be ops.aten.copy_.default

avikchaudhuri replied:
what's the difference? we use `aten.copy_` by itself elsewhere in this file

tugsbayasgalan replied:
aten.copy_ is not OpOverload i think. I guess it will do some schema mathcing thing in the runtime to find the correct overload?

avikchaudhuri replied:
Gotcha

avikchaudhuri committed:
debug ival swap (#149206)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/149206

Recall that we use "ivals" to track intermediate values of mutations during unflattening. Previously, for each such intermediate value, we would create a hidden shared attribute that would be updated / read by respective submodules.

Unfortunately this scheme doesn't work when some but not all of those submodules are swapped out. This is because the swapped in submodules have no knowledge of these hidden attributes. Thus the submodules that are not swapped out end up reading / updating dangling state.

This PR does away with these hidden attributes. Instead, we directly read the underlying buffer or placeholder that was updated, and update those underlying buffers and placeholders in place. This makes the graphs look much closer to their eager origins.

Test Plan: added some tests, ensured existing tests pass

Reviewed By: tugsbayasgalan

Differential Revision: D71203469
